pkg/controllers/capr/bootstrap/controller.go

```diff
@@ -547,3 +578,145 @@ func (h *handler) ensureMachinePreTerminateAnnotationRemoved(bootstrap *rkev1.RK
 	}
 	return bootstrap, err
 }
+
+// reconcileMachinePreDrainAnnotation reconciles the machine object that owns the bootstrap. It only reconciles the machine if it is an
+// etcd machine. Its primary purpose is to manage the pre-drain.delete.hook.machine.x-k8s.io annotation on the machine
+// object, which is used to prevent draining of the corresponding downstream node, since draining may include the static
+// etcd pod which could cause a quorum loss or at worst inability to elect a new etcd member.
+// The pre-drain hook will be set on the machine object if the machine and bootstrap are not deleting, the corresponding
+// CAPI cluster and RKEControlPlane are not deleting, and the force remove annotation is not set on the bootstrap.
+// The annotation will be removed from the machine to allow draining in the following cases:
+// * The machine is deleting and no machines are using this one's join-url, and the cluster has fully reconciled
+// * The bootstrap is missing the CAPI cluster label || the CAPI cluster controlPlaneRef is nil || the machine noderef is nil
+// * Any of the following: CAPI kubeconfig secret, CAPI cluster object, RKEControlPlane object are not found
+//
+// Notably, CAPI controllers do not trigger a deletion of the RKEBootstrap object if a pre-drain annotation exists on the corresponding machine object.
+// This means we rely on the OnChange handler to perform node safe removal, when it sees that the corresponding machine is deleting.
+//
+// Lastly, much of the logic here is shared with reconcileMachinePreTerminateAnnotation, as they have many of the same
+// prerequisite checks, however we must ensure a new init node is completely functional and all nodes have joined it
+// before draining, as otherwise etcd safe member removal may not succeed.
+func (h *handler) reconcileMachinePreDrainAnnotation(bootstrap *rkev1.RKEBootstrap) (*rkev1.RKEBootstrap, error) {
```

Comment:
Seeing as this function and the one below are largely duplicates of their `pre-terminate` counterparts, do you think it would be worthwhile to pull their common code into helper functions that they can both call?